Loop Contracts Annotation for While-Loop

qinheping opened the pull request:
This PR introduce the loop contracts annotation for while-loops using proc-macro.
A while loop of the form 
 ``` rust
#[kani::loop_invariant(inv)]
while guard {
    body
}
```
is annotated as 
 ``` rust
#[inline(never)]
#[kanitool::fn_marker = "kani_register_loop_contract"]
const fn kani_register_loop_contract_id<T, F: FnOnce() -> T>(f: F) -> T {
    unreachable!()
}
while 
kani_register_loop_contract_id(|| -> bool {inv}) && guard {
    loop_body;
}
```
We then replace the function body of the register function `kani_register_loop_contract_id` by a call to its function argument `f`. 

In the loop-contract transformation, we move the calls to register functions to a new basic block as new loop latches and redirect all loop latches (for loops containing `continue` there can be multiple latches pointing to the same loop head) to the new loop latches to make sure that
1. each loop contain only one loop latch;
2. the terminator of the loop latch is a call to the register function.

In detail, we transform
```ignore
loop_head_block: {
     loop_head_stmts
     _v = kani_register_loop_contract(move args) -> [return: next_idx];
}

 ...
loop_body_blocks
 ...

ori_loop_latch_block: {
     loop_latch_stmts
     goto -> loop_head_block;
}
```
to blocks
```ignore
// loop head block
loop_head_block: {
     _v = true
     goto -> next_idx
}

...
loop_body_blocks
...

ori_loop_latch_block: {
     loop_latch_stmts
     goto -> new_loop_latch_block;
}

new_loop_latch_block: {
     _v = kani_register_loop_contract(move args) -> [return: next_idx];
}
```
The register functions will be transformed to
```ignore
    bb0: {
        _0 = closure@fn as std::ops::FnOnce::call_once(move _1, ()) -> [return: bb1, unwind unreachable];
    }
    bb1: {
        return;
    }
```

At the end, the call to the register function will be codegened as backward goto with loop contracts annotated.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented on library/kani/src/lib.rs:
```diff
@@ -140,6 +140,25 @@ pub const fn assert(cond: bool, msg: &'static str) {
+/// that we want to annotate to.
+#[inline(never)]
+#[rustc_diagnostic_item = "KaniLoopInvariant"]
+pub const fn kani_loop_invariant(_cond: bool) {}
```
These functions seemed to be used as markers, so maybe rename that so it is explicit. This one for example seems to be marking the end of the loop, so it could be something like.
```rust
pub const fn kani_loop_end_marker() {}
```

celinval reviewed (commented):
I think I'm missing the big picture here, since the PR itself is currently a no-op and the attribute itself has no documentation.

qinheping commented on library/kani_macros/src/lib.rs:
```diff
@@ -346,6 +358,77 @@ mod sysroot {
+                Span::call_site(),
+            );
+            quote!(
+                let #closre_name = ||(#inv_expr);
+                kani::kani_loop_invariant_begin();
+                #closre_name();
+                kani::kani_loop_invariant_end();
+                #loop_stmt
```
Updated the PR description to make them consistent.

adpaco-aws commented on tests/ui/loop_contracts/small_slice_eq/small_slcie_eq.rs:
```diff
@@ -0,0 +1,41 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// kani-flags: -Z loop-contracts --keep-temps --enable-unstable --cbmc-args --arrays-uf-always --no-standard-checks
```
Do you need `--keep-temps`? Do you need `--no-standard-checks`?

adpaco-aws commented on tests/ui/loop_contracts/small_slice_eq/small_slcie_eq.rs:
```diff
@@ -0,0 +1,41 @@
+
+// kani-flags: -Z loop-contracts --keep-temps --enable-unstable --cbmc-args --arrays-uf-always --no-standard-checks
+
+// Check if loop contracts are correctly applied.
```
Please add module-level documentation with `//!` and a more concrete description of what the test is checking.

adpaco-aws commented on tests/ui/loop_contracts/small_slice_eq/small_slcie_eq.rs:
Note the typo in `slcie`.

adpaco-aws reviewed (commented):
It's looking good but I'm also missing the big picture when it comes to
 * the general approach to loop invariants - are we planning on doing the same for `for` loops?
 * the code generation for invariant/latch blocks - do you describe this transformation in the RFC? 
 * postprocessing of reachability checks - not sure what's being done or why it's needed

adpaco-aws commented on kani-compiler/src/codegen_cprover_gotoc/context/loop_contracts_ctx.rs:
```diff
@@ -0,0 +1,140 @@
+///     of the next backward jumping we codegen.
+///     We enter this state when codegen for `KaniLoopInvariantEnd`.
+///     We exit this state when codegen for the first backward jumping.
+#[allow(dead_code)]
```
Why is this needed?

adpaco-aws commented on tests/ui/loop_contracts/small_slice_eq/small_slcie_eq.rs:
```diff
@@ -0,0 +1,41 @@
+    unsafe {
+        let (mut px, mut py) = (x.as_ptr(), y.as_ptr());
+        let (pxend, pyend) = (px.add(x.len() - 4), py.add(y.len() - 4));
+        #[kani::loop_invariant( px as isize >= x.as_ptr() as isize
+        && py as isize >= y.as_ptr() as isize
+        && px as isize - x.as_ptr() as isize == (py as isize - y.as_ptr() as isize))]
```
Can you please add a simpler test with a single variable that doesn't use pointer arithmetic? Or is that not possible for some reason?

qinheping commented on cprover_bindings/src/goto_program/stmt.rs:
```diff
@@ -283,7 +286,18 @@ impl Stmt {
+    }
+
+    /// `goto dest;` with loop invariant
+    pub fn goto_with_loop_inv<T: Into<InternedString>>(
```
Good idea. Updated the code.

qinheping commented:
@celinval  @feliperodri , I refactored the implementation with closure. It is ready for review now.

celinval commented:
> @celinval @feliperodri , I refactored the implementation with closure. It is ready for review now.

Can you please update the description of the PR with an overview of the approach you implemented? Thanks

celinval commented on kani-compiler/src/codegen_cprover_gotoc/overrides/loop_contracts_hooks.rs:
```diff
@@ -0,0 +1,37 @@
+
+pub struct LoopInvariantRegister;
+
+impl GotocHook for LoopInvariantRegister {
```
tbh, I would just keep it in the same file as the other hooks. But not a big deal, can you document what this is expanding to? Thanks!

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,266 @@
+/// kani_register_loop_contract(kani_loop_invariant)
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(||->bool{true});
```
wait, why do you register a dummy closure down here?

I think we want to redefine the closure so it captures the arguments again and don't violate the borrow checker. Something like:

```rust
let kani_loop_invariant_entry = || -> bool {inv};
kani_register_loop_contract(kani_loop_invariant_entry)
while guard {
    loop_body;
    let kani_loop_invariant_latch = || -> bool {inv};
    kani_register_loop_contract(kani_loop_invariant_latch);
}
```

qinheping replied:
The idea of this implementation (and also the loop contracts in CBMC) is that we want to use one single expression as the loop invariants, and let CBMC do the transformation:  rewrite loop to if, duplicate the loop invariant (check and induction assumption) to the entry of the loop as base case, instrument havocing instructions after the induction assumption, and check the loop invariant at the end of the loop.

The one single loop invariant expression in my mind should be `kani_loop_invariant` which is `|| -> bool {inv}`. As it will be called before the entry of the loop after the transformation, it should be assigned before the loop. However, without dummy call, expanding the macro in the following way will violate the borrow check.
```
let kani_loop_invariant = || -> bool {inv};
while guard {
    loop_body;
    kani_register_loop_contract(kani_loop_invariant);
}
```

So I choose to use a dummy call as a place holder and later replace it with the actual call during MIR transform.

celinval replied:
But this MIR transformation is creating an invalid MIR that does not respect borrow checking rules.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
````diff
@@ -0,0 +1,266 @@
+///    to
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
````
don't we need to check the invariant before the loop entry?

qinheping replied:
We annotate the loop invariants in the loop latch goto and let CBMC instrument the checks.

celinval commented on library/kani_macros/src/sysroot/loop_contracts/mod.rs:
```diff
@@ -0,0 +1,101 @@
+use syn::spanned::Spanned;
+use syn::{Expr, Stmt};
+
+fn generate_unique_id_from_span(stmt: &Stmt) -> String {
```
nit: I recommend to move the private functions to the bottom of the file for readability sake.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,266 @@
+/// let kani_loop_invariant = || -> bool {inv};
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(kani_loop_invariant);
```
This could break the aliasing and ownership model.

qinheping replied:
I understand that it will violate the borrow check in Rust. But is it also a problem if we do it in MIR level?

celinval replied:
Yes. The borrow checker is trying to prevent users from making mistakes since mutation may invalidate memory.

Here is an example where I think this won't work:

```rust
type Data = u8;

struct Container {
    data: Box<Data>,
}

impl Container {
    fn new_data(&mut self, data: Data) {
        self.data = Box::new(data)
    }
}

/// Dummy function that reallocates data inside the loop.
pub fn target() {
    let mut container = Container { data: Box::new(1) };
    let mut i = 0;
    // This captures the pointer to data that will be deallocated
    let invariant = || {
        *container.data > i
    };
    kani_register_loop_contract(invariant);
    while i < 10 {
        i += 1;
        container.new_data(*container.data + i);
        let place_holder = || { true };
        kani_register_loop_contract(place_holder);
    }
}

const fn kani_register_loop_contract<T, F: FnOnce() -> T>(f: F) -> T {
    unreachable!()
}
```

I believe the transformation from this PR would trigger an invalid read by looking at the MIR generated using [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=458da18d5b304ebcf38ddec1224a5884).

It looks like the `invariant` closure captures a pointer to the allocated data (not a reference to the box):
```asm
_1 = Container { data: move _2 };
_3 = const 0_u8;
_18 = deref_copy (_1.0: std::boxed::Box<u8>);
_20 = (((_18.0: std::ptr::Unique<u8>).0: std::ptr::NonNull<u8>).0: *const u8);
_5 = &(*_20);
_6 = &_3;
_4 = {closure@[src/lib.rs:18:21: 18:23](..)} { container: move _5, i: move _6 };
```

Thus, invoking this closure in the latch (after `new_data()` call that reallocates the box) will trigger an invalid memory read.

Can you please check if that's the case?

qinheping replied:
Good point. I checked the case and it does cause problem in current transformation. I am thinking about not using closure to capture the invariants, and rolling back to the marker functions idea. I will try to refactor my code and update here.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
````diff
@@ -0,0 +1,266 @@
+///
+/// ```
+///
+/// 3. Move the call to the register function to the loop latch terminator. This is required
````
Oh... maybe clarify that we need the call to be latch's terminator.

qinheping replied:
updated

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,266 @@
+    where
+        Self: Sized,
+    {
+        true
```
What happens if the loop invariant check hasn't been enabled?

qinheping replied:
It should be true only when loop contracts is enable. Updated.

celinval commented on kani-driver/src/call_goto_instrument.rs:
```diff
@@ -183,6 +192,18 @@ impl KaniSession {
+        let args: Vec<OsString> = vec![
+            "--apply-loop-contracts".into(),
+            "--loop-contracts-no-unwind".into(),
+            "--disable-loop-contracts-side-effect-check".into(),
```
Why is that?

qinheping replied:
Because loop contracts now are wrapped in a closure which will be a side-effect expression in CBMC even they may not contain side-effect. So we disable the side-effect check for now and will implement a better check instead of simply rejecting function calls and statement expressions.
https://github.com/diffblue/cbmc/issues/8393

celinval replied:
Can you please link this issue here with your explanation?

celinval commented on kani_metadata/src/unstable.rs:
```diff
@@ -83,6 +83,8 @@ pub enum UnstableFeature {
     SourceCoverage,
     /// Enable function contracts [RFC 9](https://model-checking.github.io/kani/rfc/rfcs/0009-function-contracts.html)
     FunctionContracts,
+    /// Enable loop contracts [RFC 12]
```
Please add the link.

BTW, will this enable the entire flow where a loop is checked and later replaced?

qinheping replied:
Yes, this will enable the entire flow. The two test cases run with loop contracts enable, where the running time is only ~1 seconds comparing to timeout when running without loop contracts

celinval commented on library/kani_macros/src/sysroot/loop_contracts/mod.rs:
```diff
@@ -0,0 +1,101 @@
+                .unwrap();
+                ew.body.stmts.push(inv_end_stmt);
+            }
+            _ => (),
```
Shouldn't this also print an error?

celinval commented on tests/ui/loop_contracts/small_slice_eq/small_slice_eq.rs:
```diff
@@ -0,0 +1,44 @@
+
+//! Check if loop contracts are correctly applied. The flag --no-standard-checks should be
+//! removed once same_object predicate is supported in loop contracts.
+//! This function is originally the std function
+//! https://github.com/rust-lang/rust/blob/master/library/core/src/str/pattern.rs#L1885
```
I think we might want to use the modifications copyright here.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
This transformation seems very CBMC specific. I wonder if we should make this more explicit here. @zhassan-aws any thoughts on how would loop invariant be handled with the Aeneas backend?

qinheping replied:
Yes, the transformer is specified for CBMC loop contracts. But we can build new transformers for other backend if they are incompatible with this one.

celinval reviewed (commented):
This is awesome! Looking forward to merging this!

qinheping commented:
@celinval 
I modified the description and addressed most of the comments. The one I am not sure about is will the MIR transform break  aliasing and ownership model?

Thank you!

celinval commented on library/kani_macros/src/sysroot/loop_contracts/mod.rs:
```diff
@@ -0,0 +1,101 @@
+        // This function gets replaced by `kani::internal::call_closure`.
+        #[inline(never)]
+        #[kanitool::fn_marker = "kani_register_loop_contract"]
+        const fn #register_ident<T, F: FnOnce() -> T>(f: F) -> T {
+            unreachable!()
+        }
```
```suggestion
        const fn #register_ident<F: FnOnce() -> bool>(_f: F) -> bool {
            unreachable!()
        }
```
BTW, I believe that if you take `f` by reference, you are able to return `true` here and fix #3586.

qinheping replied:
I got the error `error[E0493]: destructor of `F` cannot be evaluated at compile-time`.
I think we can use the same idea of function contracts here that copy the loops when expanding the macro
```
let kani_loop_contract_mode = kani::internal::ORIGINAL
match kani_loop_contract_mode {
   kani::internal::loop_contract_check => #loop_with_loop_contracts_macro
   _ => #original_loop
}
```
And later modify the value of `kani_loop_contract_mode` during transformation.

But I plan to implement this after this PR.

celinval replied:
What I'm saying is that the error will go away if you change it to a reference though. Something like:
```rust
        const fn #register_ident<F: FnOnce() -> bool>(_f: &F) -> bool {
            true
        }
```

qinheping replied:
Yes, I also added two tests cases showing harness can be correctly proved when loop contracts disabled.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+    loop_contracts_enabled: bool,
+}
+
+impl GlobalPass for LoopContractPass {
```
why is this a global pass? I believe this should be a `TransformPass` instead since the transformation is local to each function.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+                }
+            }
+        } else {
+            for instance in &instances {
+                let body = instance.body().unwrap();
+                let (modified, new_body) =
+                    self.remove_register_function_calls(tcx, body, *instance);
+                if modified {
+                    transformer.cache.entry(*instance).and_modify(|transformation_result| {
+                        *transformation_result = TransformationResult::Modified(new_body);
+                    });
+                }
+            }
```
If you modify the body of `kani_register_loop_contract` to just return true, you should be able to remove this since the function is a no-op.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+    ) {
+        if self.loop_contracts_enabled {
+            // First transform functions with loop contracts.
+            for instance in &instances {
+                let body = instance.body().unwrap();
+                let (modified, new_body) = self.transform_main_body(tcx, body, *instance);
+                if modified {
+                    transformer.cache.entry(*instance).and_modify(|transformation_result| {
+                        *transformation_result = TransformationResult::Modified(new_body);
+                    });
+                }
+            }
```
Please use the `TransformPass` and remove this code.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+            }
+
+            // Move loop head statements with loop contracts to the corresponding register functions.
+            for instance in &instances {
+                let body = instance.body().unwrap();
+                let (modified, new_body) = self.transform_register_body(tcx, body, *instance);
+                if modified {
+                    transformer.cache.entry(*instance).and_modify(|transformation_result| {
+                        *transformation_result = TransformationResult::Modified(new_body);
+                    });
+                }
+            }
```
ditto

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+    /// The bb_idx of the new loop latch block.
+    /// Keys are the original loop head.
+    new_loop_latches: HashMap<usize, usize>,
+    /// Statements of loop head with loop contracts.
+    registered_stmts: HashMap<DefId, Vec<Statement>>,
```
What is this for? Why are we caching statements?

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+            }
+        } else {
+            for instance in &instances {
+                let body = instance.body().unwrap();
```
We shouldn't be using `instance.body()` since they don't contain changes from other transformation passes. This should also go away by using `TransformPass`.

celinval commented on kani-compiler/src/kani_middle/transform/body.rs:
```diff
@@ -428,6 +428,15 @@ impl MutableBody {
+    /// Replace statements from the given basic block
+    pub fn replace_statements(
+        &mut self,
+        source_instruction: &SourceInstruction,
```
Please pass the basic block instead. The source instruction is that the function adjusts the instruction so it still points to the same instruction as before.

This code is replacing the basic block code which invalidates the source instruction.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
````diff
@@ -0,0 +1,467 @@
+///    }
+///    ```
+///
+/// 3. Move the statements `loop_head_stmts` of loop head  with contracts to the body of register
+///    functions for later codegen them as statement_expression in CBMC loop contracts.
````
Sorry, how does this work? What about the local declarations? How do we know this is sound?

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Transform loops with contracts from
```
why do we need this?

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,467 @@
+use super::GlobalPass;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
```
Do we even need to invoke the loop contract at this point? I thought CBMC would handle this.

celinval reviewed (commented):
Sorry, might need to sync about this one offline. I'm not sure I understand your transformation

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,313 @@
+    run_contract_fn: Option<FnDef>,
+    /// The map from original loop head to the new loop latch.
+    /// We use this map to redirect all original loop latches to a new single loop latch.
+    new_loop_latches: HashMap<usize, usize>,
```
You can't cache this here. This object will get re-used for all body transformations. If you need to cache something that is relevant to a specific transformation, you should create a new object that only lives during the transformation of a function.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,313 @@
+        query_db.args().unstable_features.contains(&"loop-contracts".to_string())
+    }
+
+    /// Run a transformation pass on the whole codegen unit.
```
please document what this is actually doing.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,313 @@
+    /// Run a transformation pass on the whole codegen unit.
+    fn transform(&mut self, tcx: TyCtxt, body: Body, instance: Instance) -> (bool, Body) {
+        match instance.ty().kind().rigid().unwrap() {
+            RigidTy::FnDef(_func, args) => {
```
What about closures (and co-routines)... won't they be `ClosureDef`?

qinheping replied:
Good points. I think the current implementation also support loop contracts in closures and co-routines. I will enable them and add tests for them. Tracked in #3599.

celinval commented on kani-compiler/src/kani_middle/transform/loop_contracts.rs:
```diff
@@ -0,0 +1,313 @@
+                    == Some(Symbol::intern("kani_register_loop_contract"))
+                {
+                    // Replace the body of the register function with `run_contract_fn`'s.
+                    let run = Instance::resolve(self.run_contract_fn.unwrap(), args).unwrap();
+                    (true, run.body().unwrap())
```
I believe this is redundant today.

celinval commented on tests/expected/loop-contract/box.rs:
This should be a fixme test. Please also create an issue and add a link to it in the comments.

qinheping replied:
Added.

celinval reviewed (approved):
As we discussed offline, I am OK merging this just to unblock experimentation. This transformation is not sound since you are drastically moving where variables are used, and we cannot guarantee that the code does not violate the borrow checker (hence the fixme test).

Can you also please add a few more checks, including:
1. Function with multiple loops with annotations
2. Function with nested loops
3. Harness that invokes different functions that have loop contract in them
4. Test that has a `loop_invariant` but that does not pass `-Z loop-contracts`

Thanks
